### PR TITLE
Fix #356 - expand idle timeouts

### LINKED ISSUE #356: Dropped connections can be difficult to detect - expand idle timeouts
When one end of a WebSocket connection drops off the network (e.g. unplug network cable, WiFi drops out, etc.) it can be difficult for the other end to detect this. This can be particularly difficult when both ends are sending small, regular messages. The combination of relatively large send buffers and TCP being designed to handle dropped and re-established network connections means that a dropped connection is likely to go undetected for longer than is desired.

The current Session.setMaxIdleTimeout is insufficient for this use case.

I considered an option to make the maxIdleTimeout apply to reads, writes or both. While that addresses this use case, my concern is that there could be other usage patterns where this doesn't work.

To provide maximum flexibility (with hopefully minimal implementation effort) I'd like to propose we add two new idle timeouts to give:
 - Session.setMaxIdleTimeout
 - Session.setMaxReadIdleTimeout
 - Session.setMaxWriteIdleTimeout
(and corresponding getters)

All default to zero (disabled) and all are set in milliseconds.  

For more background see https://bz.apache.org/bugzilla/show_bug.cgi?id=64644

### PR DESCRIPTION
**Pull request by markt-asf**

### CODE REVIEW

**Comment by jansupol** [comment]
This looks ok by me. I like the clarification for zero and negative values. The cons for this PR is that an extra thread would get consumed, currently a thread checks for read and write timeout, now there would be one thread for each (if read and write timeout differs). This can be an issue in an application server that defines a pool of threads and the default is very low, The customers are often used to having pools with just one or two threads, unfortunately.
@joakime WDYT?

**Comment by joakime** [comment]
Eclipse Jetty is 100% async, any blocking concepts (like what exists in the Servlet spec) are faked.
A websocket connection doesn't have a dedicated thread like this PR is proposing.
Having separate read/write timeouts is actually just introducing ridiculous scheduler requirements.
On some of our servers, with 200,000 active websocket connections, this is a new requirement that isn't possible to satisfy.

**Comment by joakime** [comment]
A connection has an overall idle timeout, and activity resets the overall timeout.
This works even in the half-closed tcp socket scenarios, and satisfies the half-closed websocket closed handshake scenarios as well.

Hell, even Java itself doesn't have this kind of split on read vs write idle timeout.

**Comment by joakime** [comment]
I also don't like what this PR implies for the behaviors on WebSocket over HTTP/2 (Which Eclipse Jetty supports) and WebSocket over HTTP/3 (Which we have prelim support for).
These also have overall idle timeout, WS+HTTP/2 with a virtual stream via a mux'd TCP connection, and WS+HTTP/3 being a stream over UDP.

File: api/client/src/main/java/jakarta/websocket/Session.java [hunk 1]
```diff
@@ -173,21 +173,63 @@
...
+     * used. The read specific ({@link #getMaxReadIdleTimeout()} and/or write specific {@link #getMaxWriteIdleTimeout()}
+     * timeouts may still apply.
      *
      * @param milliseconds the number of milliseconds.
      */
     void setMaxIdleTimeout(long milliseconds);
 
+    /**
+     * Return the number of milliseconds before this conversation may be closed by the container if no messages are
+     * received in that time. A value that is zero or negative indicates that this timeout will not be used. The general
```
Comment (joakime) [inline api/client/src/main/java/jakarta/websocket/Session.java:196]: Complete message? (all data frames?)
Partial message? (some data frames? an incomplete data frame?)
Control frame?
Reset on receive of traffic?
Or reset on delivery of data frame? or any frame? or even auto-fragmented frame?
What if the connection has higher level activity not within websocket (eg: http/2 stream specific details, or http/3 stream specific details)?

File: api/client/src/main/java/jakarta/websocket/Session.java [hunk 2]
```diff
@@ -173,21 +173,63 @@
...
+     * ({@link #getMaxReadIdleTimeout()} and/or write specific {@link #getMaxWriteIdleTimeout()} timeouts may still
+     * apply.
+     *
+     * @param milliseconds the number of milliseconds.
+     */
+    void setMaxReadIdleTimeout(long milliseconds);
+
+    /**
+     * Return the number of milliseconds before this conversation may be closed by the container if no messages are
+     * sent in that time. A value that is zero or negative indicates that this timeout will not be used. The general
```
Comment (joakime) [inline api/client/src/main/java/jakarta/websocket/Session.java:216]: Sent by what?
The application has requested a send?
The extension has enough information to send something?
The websocket connection has sent it?
The HTTP/2 stream has sent it?
The TCP connection has sent it?
The HTTP/3 stream has sent it? Do we wait for confirmation on HTTP/3 of the send?

**Comment by joakime** [comment]
There's a reason why there's no separation of read vs write idle timeout on low level Socket connections.
I'm in 100% support of fixing the language around the one idle timeout, but not in introducing the split read vs write.

**Comment by markt-asf** [comment]
I agree clarifying the language around the current idle timeout would be helpful. Since this is a WebSocket level timeout, I think it should be defined in terms of WebSocket and not any underlying protocol. How about something like:
- for read, any WebSocket data (frame or part of a frame) is received from the underlying protocol
- for write, any WebSocket data (frame or part of a frame) is written to the underlying protocol

as  a starting point?

While this clarification is useful, it does not address the original feature request from the user. If separate read/write idle timeouts are not acceptable how do you propose we meet the original requirement set out in #356?

**Comment by joakime** [comment]
Lets continue to talk in #356

**Comment by markt-asf** [comment]
Closing this PR. See #383 for the replacement.

**Pull request closed without merging**